File: library/src/main/java/wannabe/grid/RemoveHiddenGrid.java
package wannabe.grid;

import java.util.Iterator;
import wannabe.Voxel;
import wannabe.grid.iterators.HiddenRemovalIterator;

/**
 * A Grid that removes all hidden voxels -- those with neighbors to the top, bottom, left, right
 * and in front. Leaves only voxels that could be visible. See also
 * {@link AllNeighbors#isSurrounded()} for more details.
 */
public class RemoveHiddenGrid implements Grid {
  private final String name;
  private final Grid source;

  public RemoveHiddenGrid(String name, Grid source) {
    this.name = name;
    this.source = source;
  }

  @Override public Iterator<Voxel> iterator() {
    return new HiddenRemovalIterator(source);
  }

  @Override public boolean isDirty() {
    return source.isDirty();
  }

  @Override public int size() {
    // TODO this is a lie
    return source.size();
  }

  @Override public AllNeighbors neighbors(Voxel voxel) {
    // Technically not true, but in practical terms may be handy to treat these as default
    return source.neighbors(voxel);
  }

  @Override public String toString() {
    return name + " no hidden (size: " + size() + ")";
  }
}

File: library/src/main/java/wannabe/grid/RemoveHiddenGrid.kt
package wannabe.grid

import wannabe.Voxel
import wannabe.grid.iterators.hiddenRemoval

/**
 * A Grid that removes all hidden voxels -- those with neighbors to the top, bottom, left, right
 * and in above. Leaves only voxels that could be visible. See also
 * [AllNeighbors.isSurrounded] for more details.
 */
class RemoveHiddenGrid(
  private val name: String,
  private val source: Grid
) : Grid {
  override fun iterator(): Iterator<Voxel> = hiddenRemoval()

  override val isDirty: Boolean
    get() = source.isDirty

  // TODO this is a lie
  override val size: Int
    get() = source.size

  // Technically not true, but in practical terms may be handy to treat these as default
  override fun neighbors(voxel: Voxel): AllNeighbors = source.neighbors(voxel)

  override fun toString(): String {
    return "$name no hidden (size: $size)"
  }
}
